src/commands/wallet/balance.rs

```diff
@@ -3,39 +3,75 @@ use clap::Args as ClapArgs;
 use eyre::ContextCompat;
 use std::sync::Arc;
 use zksync_ethers_rs::{
-    contracts::erc20::ERC20,
+    contracts::{erc20::ERC20, l2_shared_bridge::get_l2_token_from_l1_address},
+    core::utils::{format_ether, format_units},
     providers::{Middleware, Provider},
     types::Address,
+    utils::L2_ETH_TOKEN_ADDRESS,
+    ZKMiddleware,
 };
 
 #[derive(ClapArgs)]
 pub(crate) struct Args {
     #[clap(long = "token")]
     pub token_address: Option<Address>,
+    #[clap(long = "tol2", required = false)]
+    pub l2_token: bool,
     #[clap(long = "l1", required = false)]
     pub l1: bool,
 }
 
 pub(crate) async fn run(args: Args, cfg: ZKSyncConfig) -> eyre::Result<()> {
+    let l1_provider = Provider::try_from(
+        cfg.network
+            .l1_rpc_url
+            .context("L1 RPC URL missing in config")?,
+    )?;
+    let l2_provider = Provider::try_from(cfg.network.l2_rpc_url)?;
+
     let provider = if args.l1 {
-        Provider::try_from(
-            cfg.network
-                .l1_rpc_url
-                .context("L1 RPC URL missing in config")?,
-        )?
+        l1_provider.clone()
     } else {
-        Provider::try_from(cfg.network.l2_rpc_url)?
+        l2_provider.clone()
     };
+
     let wallet_address = cfg.wallet.context("Wallet config missing")?.address;
-    let network = args.l1.then_some("L1").unwrap_or("L2");
-    if let Some(token_address) = args.token_address {
+    let network = if args.l1 { "L1" } else { "L2" };
+
+    if let Some(mut token_address) = args.token_address {
+        if args.l2_token && !args.l1 {
+            token_address = get_l2_token_from_l1_address(token_address, &l2_provider).await;
+        }
```

Comment:
I would not mutate the original token address, instead I'd:
```suggestion
    if let Some(l1_token_address) = args.token_address {
        if args.l2_token && !args.l1 {
            l2_token_address = get_l2_token_from_l1_address(token_address, &l2_provider).await;
        }
```